docs/src/test/java/docs/http/javadsl/server/HttpServerExampleDocTest.java

```diff
@@ -321,6 +322,35 @@ void discardEntityManuallyCloseConnections() {
       );
         //#discard-close-connections
       }
-    
-  
+
+  public static void gracefulTerminationExample() throws Exception {
+    //#graceful-termination
+    ActorSystem system = ActorSystem.create();
+    Materializer materializer = ActorMaterializer.create(system);
+
+
+    CompletionStage<ServerBinding> binding = Http.get(system).bindAndHandle(
+        Directives.complete("Hello world!").flow(system, materializer),
+        ConnectHttp.toHost("localhost", 8080), materializer);
+
+    ServerBinding serverBinding = binding.toCompletableFuture().get(3, TimeUnit.SECONDS);
+
+    // ...
+    // once ready to terminate the server, invoke terminate:
+    CompletionStage<HttpTerminated> onceAllConnectionsTerminated =
+        serverBinding.terminate(Duration.ofSeconds(3));
+
+    // once all connections are terminated,
+    HttpTerminated httpTerminated = onceAllConnectionsTerminated.toCompletableFuture().get();
+
+    // - you can invoke coordinated shutdown to tear down the rest of the system:
+    CoordinatedShutdown.get(system).run();
```

Comment:
I'd rather register a task to one of the early phases of `CoordinatedShutdown` and that task would call `serverBinding.terminate`.

Reply:
Not entirely sure where exactly to hook it... can do that as part of https://github.com/akka/akka-http/issues/1210 ?

Reply:
If this is just an example, I guess adding a task in the `service-unbind` phase that invokes `serverBinding.terminate` should do.

I think it's what we'd do in Play [instead of using `unbind`](https://github.com/playframework/playframework/pull/8406/files#diff-fdb7a2606b698a5146eb4e044fc6f2deR403).

Reply:
Like so then: 


```
    val binding: Future[Http.ServerBinding] =
        Http().bindAndHandle(routes, "127.0.0.1", 8080)

    // ...
    // add a hook for coordinated shutdown
    CoordinatedShutdown(system).addTask(CoordinatedShutdown.PhaseServiceUnbind, "terminate-http")(() ⇒ {
      binding.flatMap { bound ⇒
        bound.terminate(3.seconds)
      } map { _ ⇒ Done }
    })

    // terminating the system will invoke coordinated shutdown,
    // and thus also apply the "terminate-http" phase in the right order
    system.terminate()

```

Reply:
Something like https://github.com/akka/akka-http/issues/1210#issuecomment-338825745 but PhaseServiceUnbind and using terminate instead of unbind.

Perhaps not show CoordinatedShutdown at all yet.
and the blocking on onceAllConnectionsTerminated, just a comment saying that you can terminate the system when onceAllConnectionsTerminated has completed would be enough for now, and then do the proper CoordinatedShutdown

Nothing of this is blocking the release though

Reply:
> Perhaps not show CoordinatedShutdown at all yet.

Yeah, I think so. And next section will explain it properly and we'll likely hook it in by default by then anyway

Reply:
Addressing like the last @patriknw comment suggests then

Reply:
> terminating the system will invoke coordinated shutdown,

this is not correct, one would have to call `CoordinatedShutdown(system).run()`
we might adjust that in the future, see https://github.com/akka/akka/issues/25213

I wouldn't show the `system.terminate()` or  `CoordinatedShutdown(system).run()` in the same snippet as the startup, since it makes no sense to shutdown immediately after starting, it's from elsewhere.